crypto/dh/dh_key.c

```diff
@@ -327,12 +327,12 @@ static int generate_key(DH *dh)
                 goto err;
 #else
             if (dh->params.q == NULL) {
-                /* secret exponent length, must satisfy 2^(l-1) <= p */
+                /* secret exponent length, must satisfy 2^l < (p-1)/2 */
                 if (dh->length != 0
                     && dh->length >= BN_num_bits(dh->params.p))
                     goto err;
                 l = dh->length ? dh->length : BN_num_bits(dh->params.p) - 1;
-                if (!BN_priv_rand_ex(priv_key, l, BN_RAND_TOP_ONE,
```

Comment:
When dh->length is zero this ends up doing
 
````
                if (!BN_priv_rand_ex(priv_key,  BN_num_bits(dh->params.p) - 2, BN_RAND_TOP_ONE,
````

So for example if p was 255 (8 bits).. (BN_num_bits(p) - 2  = 6) which is at most a value of 63..

(p-1)/2 = 254/2 = 127

Is this -2 right in this case?

Reply:
Yes it is.  Currently we may choose the invalid private key (p-1)/2 with a
very low probability.  And all private keys x > (p-1)/2 generate the
same public key as x - (p-1)/2, so there is a certain subset of DH keys
that have two private keys that result in exactly the same (g is order q generator)
or a trivial variation thereof (g is order 2q generator).  Thus not all private keys
have the same probability to be chosen, which is not correct.
Furthermore, we have to avoid the private exponent (p-1)/2 whatever it takes,
because g^((p-1)/2) mod p == 1 if g generates a q-order group
(or p-1 if g generates a 2q-order group)

Reply:
I think we need to better document this branch.

As far as I know, we come in this case for unknown primes. We should not come here for the RFC 7919 primes. I think we assume it's a safe prime. If it was dsa parameters, we would have a q, so I'm fine assuming it's a safe prime. As far as I know, we didn't used to set the length, but now we do. So I would be fine with using the same functions as dh_builtin_genparams() to set the length, that is `(2 * ossl_ifc_ffc_compute_security_bits(BN_num_bits(dh->params.p)) + 24) / 25 * 25`

Reply:
In other words `BN_priv_rand_ex(..., 275, BN_RAND_TOP_ONE, BN_RAND_BOTTOM_ANY,...)` creates 275 bit numbers i.e., numbers in the interval of 2^274..2^275-1

Reply:
hmm, your right, my observation was more or less only that the priv_key may be >= q